src/cmd/ksh93/sh/args.c

```diff
@@ -441,25 +441,28 @@ int sh_argopts(int argc, char *argv[], void *context) {
             if (argc > 0) {
                 strsort(argv, argc, sortfn);
             } else if (np && (arp = nv_arrayptr(np)) && (args = nv_aivec(np, &bits))) {
-                struct Sort *sp;
                 char *cp;
                 int i, c, keys = 0;
-                size_t nodesize;
+
                 if (keylist) {
                     for (cp = keylist; (c = *cp); cp++) {
                         if (c == ',') keys++;
                     }
                     keys++;
                 } else {
-                    keylist = Empty;
+                    keylist = "";
                 }
                 arp->nelem = nv_aipack(arp);
                 cp = nv_name(np);
                 c = strlen(cp);
-                nodesize = sizeof(struct Node) + (keys - 1) * sizeof(Namval_t *);
-                sp = (struct Sort *)malloc(sizeof(struct Sort) + strlen(keylist) +
-                                           (sizeof(char *) + 1) * keys +
-                                           (arp->nelem + 1) * (nodesize + sizeof(void *)) + c + 3);
+                // This used to multiply by `(keys - 1)` but `keys` can be zero which means the
+                // nodesize can be less than `sizeof(struct Node)` which is obviously wrong.
+                // Whether multiplying by `keys` is correct is unclear.
+                // See issue #824.
+                size_t nodesize = sizeof(struct Node) + keys * sizeof(Namval_t *);
+                struct Sort *sp =
+                    malloc(sizeof(struct Sort) + strlen(keylist) + (sizeof(char *) + 1) * keys +
+                           (arp->nelem + 1) * (nodesize + sizeof(void *)) + c + 3);
```

Comment:
This entire size calculation is suspicious (e.g., `(sizeof(char *) + 1)`)but the final `+ 3` term especially so. Three what? As written it means three bytes but why does this need three additional bytes? There are many other places in the code with a similar pattern of allocating a huge buffer that holds many distinct structures with pointers from one to another within the malloced block.